src/Workspaces/Core/Portable/Workspace/Solution/Project.cs

```diff
@@ -227,6 +241,19 @@ public TextDocument GetAdditionalDocument(DocumentId documentId)
             return ImmutableHashMapExtensions.GetOrAdd(ref _idToAdditionalDocumentMap, documentId, s_createAdditionalDocumentFunction, this);
         }
 
+        /// <summary>
+        /// Get the analyzer config document in this project with the specified document Id.
+        /// </summary>
+        public AnalyzerConfigDocument GetAnalyzerConfigDocument(DocumentId documentId)
+        {
+            if (!ContainsAnalyzerConfigDocument(documentId))
+            {
+                return null;
+            }
+
+            return ImmutableHashMapExtensions.GetOrAdd(ref _idToAnalyzerConfigDocumentMap, documentId, s_createAnalyzerConfigDocumentFunction, this);
```

Comment:
The workspace does a red node/green node thing similar to the syntax tree model that we have. Since a Document has a property to go to the parent Project, and from a Project to a Solution, it means we can't reuse those types between snapshots. So the pattern is there are classes that end in DocumentState that _do_ share the data between snapshots, but don't have those parent pointers. We then lazily create Document (i.e. non-suffixed-with-State) objects on demand which are just thin wrappers for the underlying thing that holds onto the State and the parent pointer.